src/dotnet-uninstall/Shared/BundleInfo/Versioning/BundleVersion.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.DotNet.Tools.Uninstall.Shared.Exceptions;
+using NuGet.Versioning;
+
+namespace Microsoft.DotNet.Tools.Uninstall.Shared.BundleInfo.Versioning
+{
+    public abstract class BundleVersion : IEquatable<BundleVersion>
+    {
+        public abstract BundleType Type { get; }
+        public abstract BeforePatch BeforePatch { get; }
+
+        protected readonly SemanticVersion _semVer;
+
+        public virtual int Major => _semVer.Major;
+        public virtual int Minor => _semVer.Minor;
+        public virtual int Patch => _semVer.Patch;
+        public virtual bool IsPrerelease => _semVer.IsPrerelease;
+        public virtual MajorMinorVersion MajorMinor => new MajorMinorVersion(Major, Minor);
+
+        public BundleVersion(string value)
+        {
+            if (SemanticVersion.TryParse(value, out var version))
+            {
+                _semVer = version;
+            }
+            else
+            {
+                throw new InvalidInputVersionException(value);
+            }
+        }
+
+        public override string ToString()
+        {
+            return _semVer.ToString();
+        }
+
+        public static BundleVersion FromInput<TBundleVersion>(string value)
+            where TBundleVersion :BundleVersion
+        {
+            if (typeof(TBundleVersion).Equals(typeof(SdkVersion)))
+            {
+                return new SdkVersion(value);
+            }
+
+            if (typeof(TBundleVersion).Equals(typeof(RuntimeVersion)))
```

Comment:
This issue is that I would like to call the 1-argument constructors of `RuntimeVersion` and `SdkVersion` with generic type parameters. However generic type parameters only supports 0-argument constructors. See https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters.

I think what I want is an interface that declares static classes or constructors. Interfaces for static methods are available in C# preview. However we are not using preview :(

Reply:
This method is basically a static method so it can be placed anywhere. I just find placing it in `BundleVersion` is most suitable.

Reply:
We could probably fix this when the static method interface gets into a stable C# version.

Reply:
At the same time, what benefit will you get by using the fancy type system? I think just make another constructor for RuntimeVersion will be fine. It is even easier to discover

Reply:
From my point of view, such a class inheritance hierarchy will largely reduce code reuse. I don't think another constructor for `RuntimeVersion` would help, because what I need is essentially calling a one-argument constructor or a static method when I use generic parameters. However, generic type parameters of C# can only use zero-argument constructors by using the `new()` constraint; see the link above. Hence, I suspect if we can change the current approach and reduce code reuse at the same time.